processing/src/main/java/org/apache/druid/query/scan/ScanResultValueFramesIterable.java

```diff
@@ -200,26 +228,31 @@ public FrameSignaturePair next()
       // start all the processing
       populateCursor();
       boolean firstRowWritten = false;
-      // While calling populateCursor() repeatedly, currentRowSignature might change. Therefore we store the signature
-      // with which we have written the frames
-      final RowSignature writtenSignature = currentRowSignature;
+
       FrameWriterFactory frameWriterFactory = FrameWriters.makeFrameWriterFactory(
           FrameType.COLUMNAR,
           memoryAllocatorFactory,
-          currentRowSignature,
+          currentOutputRowSignature,
           Collections.emptyList()
       );
       Frame frame;
-      try (final FrameWriter frameWriter = frameWriterFactory.newFrameWriter(new SettableCursorColumnSelectorFactory(
-          () -> currentCursor,
-          currentRowSignature
-      ))) {
+      try (final FrameWriter frameWriter = frameWriterFactory.newFrameWriter(
+          new SettableCursorColumnSelectorFactory(() -> currentCursor, currentInputRowSignature))) {
         while (populateCursor()) { // Do till we don't have any more rows, or the next row isn't compatible with the current row
           if (!frameWriter.addSelection()) { // Add the cursor's row to the frame, till the frame is full
             break;
           }
+
+          for (Integer columnNumber : nullTypedColumns) {
+            if (currentRows.get(currentRowIndex)[columnNumber] != null) {
+              throw DruidException.defensive("Expected a null value");
+            }
+          }
+
           firstRowWritten = true;
+          // Check that the columns with the null types are actually null before advancing
```

Comment:
note: isn't this comment misplaced? (note: this detail is not necessary - but it could live as an apidoc of the `validateRow` if that would be around)

Reply:
Cleaned up the code